### PR TITLE
XDS Connection closure smart logging

### LINKED ISSUE #14591: Reduce `warning - grpc Config Stream closed: 14` noise
When an XDS server terminates connection to Envoy, Envoy logs at a warning level. 

https://github.com/envoyproxy/envoy/blob/28e8d7711b91d6fa3b76827441d4285a6cd75717/source/common/config/grpc_stream.h#L100

This results in an error message like the following.  `14 == Unavailable`
```
warning envoy config StreamAggregatedResources gRPC config stream closed: 14, no healthy upstream
```
or  `5 == NotFound`
```
warning envoy config StreamAggregatedResources gRPC config stream closed: 5, Traffic Director configuration was not found for network "VPC_NAME" in project "PROJECT_NUMBER".
..
warning envoy config StreamAggregatedResources gRPC config stream closed: 5, Requested entity was not found.
```
Problem
At least some of the statuses (14, Unavailable) are retryable and *do not* require user intervention unless there are repeated failures. Management servers (like istiod) issue a reset every 10-30 mins to get even load distribution, and it causes a warning message to be emitted in envoy logs. In the vast majority of the cases this warning is not actionable.

We have seen customers alarmed by needless "Warnings" in proxy logs. Customers take warnings seriously.

Possible solutions
1. Drop the disconnect down to `INFO`, so that when users run proxy with `WARN` they are not alarmed by these messages.
2. Be more sophisticated.  `grpc_stream` already uses backoff_strategy, which means that if *after* a full back off cycle Envoy  cannot connect, then it *should* be a warning. This also means that the status and message captured on the first failure needs to be stored along with its timestamp. Emit all that info if the error persists *after* a backoff cycle.
3. Classify grpc status, some could always be warnings and some should follow (1)/(2).

@kyessenov @PiotrSikora 

Example output from istio-proxy:1.9-pre
```
2021-01-08T23:31:25.648672Z	info	sds	resource:default pushed key/cert pair to proxy
2021-01-08T23:31:25.648717Z	info	sds	Dynamic push for secret default
2021-01-08T23:37:44.608472Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T00:09:59.313367Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T00:41:35.746643Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T01:08:54.651912Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T01:41:32.345960Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T02:09:29.151835Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T02:37:37.909292Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T03:06:46.651828Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T03:39:07.444015Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T04:12:15.233814Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T04:37:55.879321Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
2021-01-09T04:37:56.188695Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 14, upstream connect error or disconnect/reset before headers. reset reason: connection failure
2021-01-09T04:37:57.703562Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 14, upstream connect error or disconnect/reset before headers. reset reason: local reset
2021-01-09T04:37:59.232159Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 14, upstream connect error or disconnect/reset before headers. reset reason: local reset
2021-01-09T04:38:02.711652Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 14, upstream connect error or disconnect/reset before headers. reset reason: local reset
2021-01-09T05:06:05.443284Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 13,
```
istio-proxy-1.9-pre
```
2021-01-10T02:39:56.243603Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 0,
2021-01-10T02:39:56.536558Z	info	xdsproxy	Envoy ADS stream established
2021-01-10T02:39:56.536715Z	info	xdsproxy	connecting to upstream XDS server: istiod-latest.istio-system.svc:15012
2021-01-10T03:08:27.842505Z	info	xdsproxy	disconnected from XDS server: istiod-latest.istio-system.svc:15012
2021-01-10T03:08:27.842644Z	warn	xdsproxy	unregister stream
2021-01-10T03:08:27.843154Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 0,
2021-01-10T03:08:28.186165Z	info	xdsproxy	Envoy ADS stream established
2021-01-10T03:08:28.186337Z	info	xdsproxy	connecting to upstream XDS server: istiod-latest.istio-system.svc:15012
2021-01-10T03:36:42.072427Z	info	xdsproxy	disconnected from XDS server: istiod-latest.istio-system.svc:15012
2021-01-10T03:36:42.072524Z	warn	xdsproxy	unregister stream
2021-01-10T03:36:42.073011Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 0,
```

### PR DESCRIPTION
**Pull request by mandarjog**
XDS connection closures are logged as warnings only for
repeated failures for certain type of status codes.

Fixes #14591 

Signed-off-by: Mandar U Jog <mjog@google.com>

Will follow up with tests if this approach is ok.

### CODE REVIEW

File: source/common/config/grpc_stream.h [hunk 1]
```diff
@@ -15,6 +15,18 @@
...
+namespace {
+
+constexpr auto CloseLogMessage = "{} gRPC config stream closed: {}, {}";
+constexpr auto CloseLogMessageWithMs = "{} gRPC config stream closed {}ms ago: {}, {}";
+
+// TODO(htuch): Make this configurable.
+constexpr uint32_t RetryInitialDelayMs = 500;
+constexpr uint32_t RetryMaxDelayMs = 30000; // Do not cross more than 30s
+
+constexpr int64_t UnsetStatus = -9999;
```
Comment (htuch) [inline source/common/config/grpc_stream.h:27]: Maybe use `absl::optional`?

File: source/common/config/grpc_stream.h [hunk 2]
```diff
@@ -60,12 +69,13 @@ class GrpcStream : public Grpc::AsyncStreamCallbacks<ResponseProto>,
     }
     stream_ = async_client_->start(service_method_, *this, Http::AsyncClient::StreamOptions());
     if (stream_ == nullptr) {
-      ENVOY_LOG(warn, "Unable to establish new stream");
+      ENVOY_LOG(debug, "Unable to establish new grpc config stream");
```
Comment (htuch) [inline source/common/config/grpc_stream.h:72]: Shouldn't this be a warning? There's a real failure ocurring.

File: source/common/config/grpc_stream.h [hunk 3]
```diff
@@ -85,6 +95,7 @@ class GrpcStream : public Grpc::AsyncStreamCallbacks<ResponseProto>,
   void onReceiveMessage(ResponseProtoPtr<ResponseProto>&& message) override {
     // Reset here so that it starts with fresh backoff interval on next disconnect.
     backoff_strategy_->reset();
+    unsetFailure();
```
Comment (htuch) [inline source/common/config/grpc_stream.h:98]: Why is this needed? Isn't it sufficient to unset on successful connect?

**Review by htuch — CHANGES_REQUESTED** [review]
Thanks; @mandarjog can you add tests for all these new behaviors? There's a lot of different branch conditions, so we should have some coverage of the logic here.
/wait

File: source/common/config/grpc_stream.h (hunk 1 above)
Comment (mandarjog) [reply source/common/config/grpc_stream.h:27]: ack.

File: source/common/config/grpc_stream.h (hunk 2 above)
Comment (mandarjog) [reply source/common/config/grpc_stream.h:72]: The real failure with error message is recorded in the call back. I don;t think there is a case where something is not recorded in the callback, but *is* recorded here. This is also the source of noise since it does not tell the reason.

**Comment by howardjohn** [comment]
A release note may be nice, as this will impact everyone's logs

File: source/common/config/grpc_stream.h (hunk 3 above)
Comment (mandarjog) [reply source/common/config/grpc_stream.h:98]: It is the comment on the next line that lead me to believe I need a reset here, to.
But I think it makes sense to remove it. ack.

**Review by howardjohn — COMMENTED** [review]
I am not sure masking errors to debug level is the right approach.

Moving status code 0 to debug level makes sense to me, but I think we want to see errors?

**Comment by howardjohn** [comment]
I wasn't sure from the code, do we log the first error?

```
XDS Failed 0
XDS Failed 1
XDS Failed 2
XDS Failed 3
XDS connected
```

For example, in the above ^, do we log the failure 3 or failure 0?

In my opinion, logging the first failure is ideal, because if we have a log on first failure and successful connections, then we know that everything between failure 0 and "connected" is more failures, even if they are not explicitly logged. if we log only latter failures we may miss when the issues first start.

**Comment by mandarjog** [comment]
@htuch can you point me to an example of faking time  ?

**Comment by mandarjog** [comment]
If xds reconnected in the allotted time, then we do not log anything. ( there is no connection successful message yet)

at debug level, we log every event. 

If the connection does not recover in allotted time, we will print the original error and state the amount of time this condition has been going on.

**Comment by htuch** [comment]
@mandarjog you want a fake time source for the dispatcher, see the `MockDispatcher` `GlobalTimeSystem`. You probably want to return a `SimulatedTimeSystem` or `MockTimeSystem` instead.

**Comment by mandarjog** [comment]
precheck deps succeeds locally `bazel query 'deps(//source/exe:envoy_main_common_with_core_extensions_lib)'` for me.

**Comment by mandarjog** [comment]
I am thinking of adding and INFO message on successful connection, and it will also state how long was it disconnected.
`onReceiveMessage` seems like a positive sign that we are connected to the controlplane, more so than 
` async_client_->start(service_method_, *this, Http::AsyncClient::StreamOptions()) != nullptr`.  @PiotrSikora @htuch ?

**Comment by htuch** [comment]
Sounds reasonable, this is a pretty important diagnostic that applies to many use cases. We can always drop log level if folks find it too spammy.
